자판기 2단계 - 음료수 객체 리팩토링

popsmile opened the pull request:
- `Milk` , `Coffee` , `SoftDrink` 클래스를 각각 상속받는 하위 클래스를 생성했습니다.
- 위 3개의 클래스가 각각 상위 `Beverage` 클래스 외의 속성 및 메소드를 갖도록 추가했습니다.
- 하위 클래스를 상위 클래스 속성 및 메소드로 잘 동작하는지 단위테스트를 통해 다형성을 검사했습니다.

popsmile committed:
Add necessary files

popsmile committed:
Add an extension of Date by using addingTimeInterval()

popsmile committed:
Create new sub-classes inherited by Coffee

popsmile committed:
Create an enum BeveragePackage

popsmile committed:
Add an own property of the class SoftDrink

Create new sub-classes inherited by SoftDrink.

popsmile committed:
Add some own properties and methods to the class Milk

popsmile committed:
Create new sub-classes inherited by Milk

popsmile committed:
Add a super class to the class Beverage

Change the protocol CustomStringConvertible to the inheritance of NSObject to use the property named className

popsmile committed:
Override a property named className of NSObject

popsmile committed:
Add another convenience initializer to sub-classes of Milk

popsmile committed:
Create a Unit Test Bundle to test polymorphism of the classes

godrm commented on VendingMachine/VendingMachine/main.swift:
```diff
@@ -10,13 +10,13 @@ import Foundation
-    let cantata = Coffee(brand: "롯데칠성", name: "칸타타", volume: 275, price: 2200, dateOfManufacture: Date(timeIntervalSinceNow: -20*aDay))
+    let chocolateMilk = ChocolateMilk(brand: "남양유업", name: "초코에몽", volume: 180, price: 1000, dateOfManufacture: Date(timeIntervalSinceNow: -2*aDay))
+    let sprite = Sprite(volume: 250, price: 1200, dateOfManufacture: Date(timeIntervalSinceNow: -30*aDay), package: .can)
+    let cantata = Cantata(name: "칸타타 더치블랙", volume: 275, price: 2200, dateOfManufacture: Date(timeIntervalSinceNow: -20*aDay), ice: false)
```
객체를 생성할 때마다 aDay를 선언해서 써야할까요? 
더 좋은 방법이 있으면 확장해서 만들어서 쓰는 것도 고려해보세요.

godrm commented on VendingMachine/VendingMachineUnitTest/UnitTestMilk.swift:
```diff
@@ -0,0 +1,83 @@
+    }
+
+    func testClassMilkAndOneSubClassOfMilkHaveSameGroup() {
+        XCTAssertEqual(pureMilk.group, strawberryMilk.group)
```
그룹인지 비교하는것도 값을 가져와서 비교하는게 아니라 해당 객체에 메소드를 만들어보세요.

popsmile committed:
Add extensions to simplify the use of the class Date

popsmile committed:
Create a class BeverageGroup inherited by Beverage

Replace the protocol Groupable by the class BeverageGroup. Make the classes named Milk, Coffee and SoftDrink inherit the method isInSameGroup(with:) from the class BeverageGroup.

popsmile committed:
Order the groups and files by their roles

popsmile commented:
- 자주 쓰이는 `Date` 객체를 보다 간단하게 활용할 수 있도록 `extension` 했습니다.
- 같은 그룹인지 비교하는 메소드 구현 내용이 `Groupable` 을 채택하고있는 클래스에서는 모두 동일하여, `BeverageGroup` 클래스를 추가해 해당 프로토콜을 대체했습니다.

godrm commented on VendingMachine/VendingMachine/BeverageGroup.swift:
```diff
@@ -0,0 +1,21 @@
+
+    func isInSameGroup(with beverage: BeverageGroup) -> Bool {
+        return self.group == beverage.group
+    }
```
이런 비교는 isEqual-을 붙이는 경우가 많기는 합니다. 참고하세요.
